agent/internal/container/container.go

```diff
@@ -0,0 +1,439 @@
+		c.finalize(ctx, c.run(ctx))
+	})
+
+	go func() {
+		c.wg.Wait()
+		close(c.done)
+	}()
+
+	return c
+}
+
+// Reattach an existing container and receive a handle to interact with it.
+func Reattach(
+	container cproto.Container,
+	cl *docker.Client,
+	pub events.Publisher[Event],
+) *Container {
+	c := &Container{
+		// TODO(Brad): We should be recovering the allocation ID for logging.
+		containerID: container.ID,
+		// We don't need the spec because we only reattach launched containers.
+		devices: container.Devices,
+		log: logrus.WithFields(logrus.Fields{
+			"component":  "container",
+			"cproto-id":  container.ID,
+			"reattached": true,
+		}),
+		docker:  cl,
+		pub:     pub,
+		state:   container.State,
+		signals: make(chan syscall.Signal, 16), // Not infinite, but large enough to not drop often.
+		done:    make(chan struct{}),
+
+		wg: waitgroupx.WithContext(context.Background()),
+	}
+
+	c.wg.Go(func(ctx context.Context) {
+		defer c.wg.Cancel()
+		c.finalize(ctx, c.reattach(ctx))
+	})
+
+	go func() {
+		c.wg.Wait()
+		close(c.done)
+	}()
+
+	return c
+}
+
+// Summary returns a snapshot of the container state.
+func (c *Container) Summary() cproto.Container {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.summary()
+}
+
+// Detach the monitoring loops without affecting the Docker container.
+func (c *Container) Detach() {
+	c.log.Trace("detach called")
+	c.wg.Cancel()
+	c.Wait()
+}
+
+// Stop the Docker container by killing it and awaiting its exit.
+func (c *Container) Stop() {
+	c.log.Trace("close called")
+	c.Signal(context.TODO(), syscall.SIGKILL)
+	c.Wait()
+}
+
+// Signal asynchronously delivers the signal. Delivery failures are surfaced in logs.
+func (c *Container) Signal(ctx context.Context, s syscall.Signal) {
+	select {
+	case c.signals <- s:
+	case <-c.done:
+		c.log.Warnf("ignoring signal on exited container: %v", s)
+	case <-ctx.Done():
+		c.log.Warnf("ignoring signal on container due to cancellation: %v", s)
+	}
+}
+
+// Wait until the container exits. Always returns a ContainerExit unless canceled by Detach.
+func (c *Container) Wait() *aproto.ContainerStateChanged {
+	<-c.done
+	return c.exit
+}
+
+// run the container. If the context is canceled, the container is detached (as is) and the context
+// error is returned. If the container is killed, the container is cleaned up and a termination with
+// a stack trace or kill result is returned. If the container succeeds, the error will be nil.
+func (c *Container) run(parent context.Context) (err error) {
+	c.log.Trace("starting container launch")
+	launchgroup := errgroupx.WithContext(parent)
+
+	c.log.Trace("kicking off goroutine shim SIGKILL to cancellations, until we have launched")
+	launchgroup.Go(func(ctx context.Context) error {
+		for {
+			select {
+			case signal := <-c.signals:
+				switch signal {
+				case syscall.SIGKILL:
+					c.log.Tracef("signal %s, canceling run-scoped context", signal)
+					launchgroup.Cancel()
+					return ErrKilledBeforeRun
```

Comment:
I like this better than what we had before. instead of having custom deferred func process "which error is it", we've got "which error did we get first" errgroup do it instead.

Reply:
great